### ISSUE TITLE
MKL not used when static linking with WHOLE_ARCHIVE

### ISSUE THREAD

**Issue by panosk** [OpenNMT/CTranslate2#1205]
I've spotted a weird issue in my Windows build that prevents using MKL. My CMakeLists.txt is like this:

```CMake
...
add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/ext/CTranslate2)
list(APPEND LIBRARIES ctranslate2)
...
add_executable(${PROJECT_NAME}...)
...
target_link_libraries(${PROJECT_NAME}
        PUBLIC "$<LINK_LIBRARY:WHOLE_ARCHIVE,ctranslate2.lib>"
        ${LIBRARIES})
```
When I check at runtime, I see that MKL is not used:
```
[info] CPU: GenuineIntel (SSE4.1=true, AVX=true, AVX2=true, AVX512=true)
[info]  - Selected ISA: AVX2
[info]  - Use Intel MKL: false
[info]  - SGEMM backend: none (packed: false)
[info]  - GEMM_S16 backend: none (packed: false)
[info]  - GEMM_S8 backend: none (packed: false, u8s8 preferred: false)
```
If I remove the `WHOLE_ARCHIVE` line, then `Use Intel MKL: true`, but the app immediately crashes as the linker strips symbols from the ctranslate2.lib. I searched a lot but can't find a solution, and for now it's not easy to try to switch to dynamic linking with dll. Any help will be greatly appreciated. 

Note: I'm using CTranslate2 v3.12.0.

**Comment by panosk** [comment]
Here are all the relevant parts:

```CMake
-- The CXX compiler identification is MSVC 19.35.32217.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.35.32215/bin/HostX64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The C compiler identification is MSVC 19.35.32217.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.35.32215/bin/HostX64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- VERSION: 0.1.96
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE  

....

-- Build type: Release
-- Compiling for multiple CPU ISA and enabling runtime dispatch
-- Using OpenMP: C:/Program Files (x86)/Intel/oneAPI/compiler/latest/windows/compiler/lib/intel64_win/libiomp5md.lib
-- Found MKL include directory: C:/Program Files (x86)/Intel/oneAPI/mkl/latest/include
-- Found MKL library directory: C:/Program Files (x86)/Intel/oneAPI/mkl/latest/lib/intel64
-- Found DNNL include directory: C:/Program Files (x86)/Intel/oneAPI/dnnl/latest/cpu_iomp/include
-- Found DNNL library: C:/Program Files (x86)/Intel/oneAPI/dnnl/latest/cpu_iomp/lib/dnnl.lib
-- NVCC host compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.35.32215/bin/HostX64/x64/cl.exe
-- NVCC compilation flags: -Xcompiler=/MT$<$<CONFIG:Debug>:d>;-gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_53,code=sm_53;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_86,code=compute_86
CMake Warning at ext/CTranslate2/CMakeLists.txt:459 (message):
  cuDNN library is not enabled: convolution layers will not be supported on
  GPU
  
  ....
```
Note that if I force the use of MKL with `CT2_USE_MKL=1`, then it is used, so it seems it is not selected by default on Intel CPUs for some reason.

**Comment by guillaumekln** [comment]
This is strange. Even if the Intel CPU is not detected for some reasons, we would expect "DNNL" to be reported in the info logs:

```
[info]  - SGEMM backend: DNNL (packed: false)
[info]  - GEMM_S16 backend: none (packed: false)
[info]  - GEMM_S8 backend: DNNL (packed: false, u8s8 preferred: true)
```

Based on your `CMakeLists.txt`, it seems ctranslate2 appears 2 times in `target_link_libraries`. Can it be an issue?

**Comment by panosk** [comment]
I apologize for the confusion. In the log in my first post, I compiled without DNNL, as I was experimenting with MKL to see what's happening. In the CMake log in my last post, I compile with DNNL, which is my regular workflow, and indeed DNNL is found and used by default when compiling both with MKL and DNNL.

> Based on your `CMakeLists.txt`, it seems ctranslate2 appears 2 times in `target_link_libraries`. Can it be an issue?

This is needed when using `WHOLE_ARCHIVE`.

**Comment by guillaumekln** [comment]
I'm not sure what's happening so I'm looking to remove the need to use "whole archive".

Can you help testing the branch #1241?

It should work without using "whole archive". You could check that the program does not crash and the logger is correctly initialized (the environment variable `CT2_VERBOSE=1` should work).

**Comment by panosk** [comment]
Yes, sure, I'll check later today, thank you.

**Comment by panosk** [comment]
It seems to work fine! I left out `WHOLE_ARCHIVE`, MKL is used by default, and model translates without crashes. With `WHOLE_ARCHIVE` I get the previous behavior and MKL is not used. I only tested on Windows --macOS on Intel didn't seem to have this issue, even with the equivalent of `WHOLE_ARCHIVE`, `-force_load`. One minor (and insignificant for me) change I noticed is that `CT2_VERBOSE=1` no longer shows batch information during translation.
